Experimental QNX Support

Hoverbear opened the pull request:
Introduces experimental QNX support.

## Open Problems

* Moving beyond QNX 7.1.0 we likely won't see Mac versions, so we should consider how to deal with that.

Hoverbear commented on .circleci/workflows.yml:
```diff
@@ -159,11 +159,12 @@ jobs:
     environment:
       FERROCENE_HOST: x86_64-unknown-linux-gnu
-      FERROCENE_TARGETS: aarch64-unknown-linux-gnu,aarch64-unknown-none,thumbv7em-none-eabi,thumbv7em-none-eabihf,armv8r-none-eabihf,wasm32-unknown-unknown,armv7r-none-eabihf,armebv7r-none-eabihf
+      FERROCENE_TARGETS: aarch64-unknown-linux-gnu,aarch64-unknown-none,thumbv7em-none-eabi,thumbv7em-none-eabihf,armv8r-none-eabihf,wasm32-unknown-unknown,armv7r-none-eabihf,armebv7r-none-eabihf,aarch64-unknown-nto-qnx710,x86_64-pc-nto-qnx710
```
Noting I still need to add the self-test bits but also that requires more circle.yaml changes since I need to get the QNX toolchain there, too.

Hoverbear commented:
bors try

pietroalbini commented on .circleci/workflows.yml:
```diff
@@ -681,6 +685,20 @@ commands:
+          steps:
+            - run:
+                name: Fetch QNX710 toolchain
+                command: aws s3 cp s3://ferrocene-ci-artifacts/qnx/qnx710.tar.zst - | zstd -d | tar x
```
Can we store this in `s3://ferrocene-ci-mirrros/manual`? That's where everything else we mirrored is stored.

pietroalbini commented on .circleci/workflows.yml:
```diff
@@ -681,6 +685,20 @@ commands:
+                command: echo 'source $(pwd)/qnx710/qnxsdp-env.sh' >> "$BASH_ENV"
+            - run:
+                name: Start QNX License Server port forwarding
+                command: aws ssm start-session --target i-0accef99ca2244097 --document-name AWS-StartPortForwardingSession --parameters '{"portNumber":["56859"], "localPortNumber":["56859"]}'
```
I'd add a tag (probably `Name`?) to the instance [and first search the instance by tag](https://serverfault.com/a/560534) to get its ID. That way, if we ever need to recreate the instance for whatever reason we don't need to go and manually change the CI configuration.

Hoverbear replied:
Given that `Describe*` style actions [can't be limited to specific resources](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-policy-structure.html#ec2-supported-iam-actions-resources) I don't want to do that since it would require permitting the CI token to enumerate all the ec2 nodes of our infrastructure. (It might seem harmless now, but I don't think it's a wise road to travel down)

Perhaps we could have terraform [provision a file to the bucket with the QNX artifacts](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_object) which gets recreated if the node changes?

pietroalbini reviewed (commented):
The permission error in CI is due to the policy don't allowing access to the whole `s3://ferrocene-ci-artifacts`, but just `s3://ferrocene-ci-artifacts/ferrocene`. I left a comment for where it should be stored.

Also, when adding the target in `packages.toml` please add it with the `qnx` subset rather than the `default` subset, as we want to gate access to it separately.

Hoverbear commented on ferrocene/tools/self-test/src/targets.rs:
```diff
@@ -22,6 +22,10 @@ static SUPPORTED_TARGETS: &[TargetSpec] = &[
     TargetSpec { triple: "armv7r-none-eabihf", std: false, linker: Linker::BundledLld },
     TargetSpec { triple: "armebv7r-none-eabihf", std: false, linker: Linker::BundledLld },
+    TargetSpec { triple: "armv7r-none-eabihf", std: false, linker: Linker::BundledLld },
+    TargetSpec { triple: "armebv7r-none-eabihf", std: false, linker: Linker::BundledLld },
```
Oops

Hoverbear commented:
bors try

Hoverbear commented:
bors try

Hoverbear commented:
bors try-

Hoverbear commented:
bors try

Hoverbear commented:
bors try

How about this? It **should** fail since you can't actually create the SSM session and thus can't get the license.

Hoverbear commented:
bors try

Hoverbear commented:
bors try

Hoverbear commented:
bors try

Hoverbear commented:
bors try

I think I fixed the permissions...

Hoverbear commented:
bors try

I knew I forgot something!

Hoverbear commented:
bors try

I can't believe it, I forgot to set QNXLM_LICENSE_FILE after all that. 😅

Hoverbear commented:
bors try

Hoverbear commented:
bors try

Hoverbear commented:
bors merge

Hoverbear commented:
bors try

pietroalbini commented on ferrocene/doc/internal-procedures/src/partners/qnx.rst:
```diff
@@ -0,0 +1,206 @@
+
+.. note::
+    
+    QNX only supports x86-64 Linux (glibc) and x86-64 Windows.
+    
+    QNX does not support Apple Silicon macOS. QNX is deprecating support
+    for x86-64 macOS.
```
Very small nit: use `:target:` to refer to the targets?

pietroalbini commented on ferrocene/doc/user-manual/src/targets/index.rst:
```diff
@@ -102,6 +102,18 @@ should not be used in production.
+     - ``aarch64-unknown-nto-qnx710``
+     - Cross-compilation
+     - Full
+     - Only available on targets supported by QNX (:ref:`x86_64-unknown-linux-gnu` and :ref:`x86_64-pc-windows-msvc`)
```
I'd write it like "Can only be cross-compiled from :ref:`x86_64-unknown-linux-gnu` and :ref:`x86_64-pc-windows-msvc`"

pietroalbini reviewed (commented):
Minor docs nits, otherwise looks good!

Hoverbear commented:
bors try

Hoverbear commented:
I updated our deployment image to remove armv7le support and the docs. That should reduce the number of files Windows struggles with.

Hoverbear commented:
bors try

Hoverbear commented:
bors try

Hoverbear commented:
bors try

Hoverbear commented:
I removed the QNX windows self-test and updated the docs to note this for now. I'll circle back around to try to fix 43+ minute tarball unpack on Windows. I'm not sure how to resolve it as it seems to be a simple case of "No IOPS"

Hoverbear commented:
bors try

pietroalbini reviewed (approved):
bors merge
